server/src/main/java/org/elasticsearch/index/mapper/MappedFieldType.java

```diff
@@ -453,6 +453,13 @@ public boolean eagerGlobalOrdinals() {
         return false;
     }
 
+    /**
+     * @return if this field type should be matched by a wildcard field expansion
+     *
+     * @see org.elasticsearch.index.search.QueryParserHelper
+     */
+    public abstract boolean includeInFieldExpansion(SearchExecutionContext context);
```

Comment:
I'm thinking of giving more constraints to the contract, e.g. it is only legal to return `false` if a term query on this field couldn't possibly match any documents?

Reply:
I think it's a bit more general than that - for example, we have internal metadata fields like the dynamically generated flattened key fields that return 'false' here, but would return term queries.  But then thinking about it these sort of fields would never be included in the expansion anyway as they're not returned by `getMatchingFieldNames()`

Reply:
It would be nice to have these subtleties in javadocs. :)

Reply:
True! And we may use this in other places in future.  I'll add a note.

Reply:
I wonder if the method should have such a broad name/ scope. It doesn't actually have the "final say" about whether a field is included in an expansion -- `QueryParserHelper` adds other constraints based on `TextSearchInfo` and whether it thinks the field is a metadata field. This method also isn't used in all field expansion contexts, for example when loading `fields`. Could this method instead be something like `fieldMayExistInIndex`?